dns-policy-operator/src/charm.py

```diff
@@ -0,0 +1,352 @@
+        """Convert DNSRecordRequirerData to a list of DnsEntry.
+
+        Args:
+            relation_data: input relation data
+        Returns:
+            A list of DnsEntry
+        """
+        entries: list[models.DnsEntry] = []
+        for record_requirer_data, _ in relation_data:
+            for entry in record_requirer_data.dns_entries:
+                entries.append(models.create_dns_entry_from_requirer_entry(entry))
+        return entries
+
+    def record_requirer_data_to_zones(
+        self,
+        record_requirer_data: dns_record.DNSRecordRequirerData,
+    ) -> list[models.Zone]:
+        """Convert DNSRecordRequirerData to zone files.
+
+        Args:
+            record_requirer_data: The input DNSRecordRequirerData
+
+        Returns:
+            A list of zones
+        """
+        zones_entries: dict[str, list[models.RequirerEntry]] = {}
+        for entry in record_requirer_data.dns_entries:
+            if entry.domain not in zones_entries:
+                zones_entries[entry.domain] = []
+            zones_entries[entry.domain].append(entry)
+
+        zones: list[models.Zone] = []
+        for domain, entries in zones_entries.items():
+            zone = models.Zone(domain=domain, entries=set())
+            for entry in entries:
+                zone.entries.add(models.create_dns_entry_from_requirer_entry(entry))
+            zones.append(zone)
+        return zones
+
+    def _get_remote_relation_data(self) -> dns_record.DNSRecordProviderData | None:
+        """Get the dns_record remote relation data.
+
+        This function is used to get performance statistics on the remote relation data retrieval.
+
+        Returns:
+            the dns_record remote relation data
+        """
+        start_time = time.time_ns()
+        relation_data = self.dns_record_provider.get_remote_relation_data()
+        logger.debug(
+            "Relation data retrieval duration (ms): %s", (time.time_ns() - start_time) / 1e6
+        )
+        return relation_data
+
+    def _on_config_changed(self, event: ops.ConfigChangedEvent) -> None:
+        """Handle changed configuration."""
+        # Fetch the new config value
+        log_level = typing.cast(str, self.model.config["log-level"]).lower()
+
+        # Do some validation of the configuration option
+        if log_level not in VALID_LOG_LEVELS:
+            # In this case, the config option is bad, so block the charm and notify the operator.
+            self.unit.status = ops.BlockedStatus(f"invalid log level: '{log_level}'")
+
+        self.unit.status = ops.MaintenanceStatus("Configuring workload")
+        self.dns_policy.configure(
+            {
+                "debug": "true" if self.config["debug"] else "false",
+                "allowed-hosts": json.dumps(
+                    [e.strip() for e in str(self.config["allowed-hosts"]).split(",")]
+                ),
+            }
+        )
+        self.unit.status = ops.ActiveStatus("")
```

Comment:
Since you're not in an holistic charm, this can mess up with previously set status, for example if the DB relation is removed and then you do a juju config the status will be back to active
If you're not using an holistic approach I suggest to use the collect_status event to compute the final status and set the status nowhere else

Reply:
That was removed in favor of collect_status()